### PR TITLE
2016/discord: Add discord auth to social auth providers

### PR DESCRIPTION
**Pull request by wardbox**
### Description

> Adds discord auth to social auth providers. Contributes to #2016 but doesn't fix.

### Select what type of change this PR introduces:
1. [ ] **Just code/docs improvement** (no functional change).
2. [ ] **Bug fix** (non-breaking change which fixes an issue).
3. [x] **New feature** (non-breaking change which adds functionality).
4. [ ] **Breaking change** (fix or feature that would cause existing functionality to not work as expected).

### Update Waspc ChangeLog and version if needed
If you did a bug fix, new feature, or breaking change, that affects waspc, make sure you satisfy the following:
1. [x] I updated [ChangeLog.md](https://github.com/wasp-lang/wasp/blob/main/waspc/ChangeLog.md) with description of the change this PR introduces.
2. [x] I bumped waspc version in [waspc.cabal](https://github.com/wasp-lang/wasp/blob/main/waspc/waspc.cabal) to reflect changes I introduced, with regards to the version of the latest wasp release, if the bump was needed.

### Update example apps if needed
If you did code changes and added a new feature or modified an existing feature, make sure you satisfy the following:
1. [x] I updated `waspc/examples/todoApp` as needed (updated modified feature or added new feature) and manually checked it works correctly.
2. [x] I updated `waspc/headless-test/examples/todoApp` and its e2e tests as needed (updated modified feature and its tests or added new feature and new tests for it).

### CODE REVIEW

**Comment by wardbox** [comment]
Heads up I did have some test failures from both `cabal test` and the headless test, but wasn't seeing a clear reason they'd be failing. Very likely due to lack of familiarity with the test suite, might need some help deciphering those!

File: waspc/ChangeLog.md [hunk 1]
```diff
@@ -1,5 +1,11 @@
 # Changelog
 
+## 0.14.1
```
Comment (wardbox) [inline waspc/ChangeLog.md:3]: Wasn't sure how the versioning system worked, but looking at it now would this go in 0.13.3 instead?

File: waspc/data/Generator/templates/sdk/wasp/auth/forms/internal/social/SocialIcons.tsx [hunk 2]
```diff
@@ -5,6 +5,22 @@ const defaultStyles = css({
...
+    viewBox="0 0 24 24"
+  >
+    <g id="brand" stroke="none" strokeWidth="1" fill="none" fillRule="evenodd">
+      <g id="discord" fill="#000000" fillRule="nonzero">
+        <path d="M18.59 5.88997C17.36 5.31997 16.05 4.89997 14.67 4.65997C14.5 4.95997 14.3 5.36997 14.17 5.69997C12.71 5.47997 11.26 5.47997 9.83001 5.69997C9.69001 5.36997 9.49001 4.95997 9.32001 4.65997C7.94001 4.89997 6.63001 5.31997 5.40001 5.88997C2.92001 9.62997 2.25001 13.28 2.58001 16.87C4.23001 18.1 5.82001 18.84 7.39001 19.33C7.78001 18.8 8.12001 18.23 8.42001 17.64C7.85001 17.43 7.31001 17.16 6.80001 16.85C6.94001 16.75 7.07001 16.64 7.20001 16.54C10.33 18 13.72 18 16.81 16.54C16.94 16.65 17.07 16.75 17.21 16.85C16.7 17.16 16.15 17.42 15.59 17.64C15.89 18.23 16.23 18.8 16.62 19.33C18.19 18.84 19.79 18.1 21.43 16.87C21.82 12.7 20.76 9.08997 18.61 5.88997H18.59ZM8.84001 14.67C7.90001 14.67 7.13001 13.8 7.13001 12.73C7.13001 11.66 7.88001 10.79 8.84001 10.79C9.80001 10.79 10.56 11.66 10.55 12.73C10.55 13.79 9.80001 14.67 8.84001 14.67ZM15.15 14.67C14.21 14.67 13.44 13.8 13.44 12.73C13.44 11.66 14.19 10.79 15.15 10.79C16.11 10.79 16.87 11.66 16.86 12.73C16.86 13.79 16.11 14.67 15.15 14.67Z">
```
Comment (wardbox) [inline waspc/data/Generator/templates/sdk/wasp/auth/forms/internal/social/SocialIcons.tsx:17]: Just grabbed a SVG from some site, not sure if y'all use a specific source for these

File: web/docs/auth/Pills.css [hunk 3]
```diff
@@ -1,6 +1,7 @@
 :root {
   --auth-pills-color: #333;
   --auth-pills-email: #e0f2fe;
+  --auth-pills-discord: #f1f5f9;
```
Comment (wardbox) [inline web/docs/auth/Pills.css:4]: Wasn't sure what color to use for these, copilot picked this one :^)

**Comment by Martinsos** [comment]
All right this is quite cool @wardbox !
@infomiho is the master of Auth, let's see what he says -> there will probably be changes needed and additions on top of this (I guess docs especially), but this is looking like a potentially good foundation!

**Comment by wardbox** [comment]
> All right this is quite cool @wardbox ! @infomiho is the master of Auth, let's see what he says -> there will probably be changes needed and additions on top of this (I guess docs especially), but this is looking like a potentially good foundation!

I actually have no idea where my doc went, must have removed in a stash 🤔 will submit that shortly

File: waspc/ChangeLog.md (hunk 1 above)
Comment (infomiho) [reply waspc/ChangeLog.md:3]: Since we will be releasing 0.14.0 in a mid-July, I believe this will get release within that version. So, write the update under the `0.14.0`.

File: waspc/headless-test/examples/todoApp/todoApp.wasp [hunk 4]
```diff
@@ -20,6 +20,7 @@ app todoApp {
           clientRoute: PasswordResetRoute
         }
       },
+      discord: {},
```
Comment (infomiho) [inline waspc/headless-test/examples/todoApp/todoApp.wasp:23]: No need to modify the headless tests to include Discord 👍 we only included the Google one to test out the social login provider React components.

I'd remove any Discord related changes in the headless tests.

File: waspc/waspc.cabal [hunk 5]
```diff
@@ -6,7 +6,7 @@ cabal-version: 2.4
 --    Consider using hpack, or maybe even hpack-dhall.
 
 name:           waspc
-version:        0.14.0
+version:        0.14.1
```
Comment (infomiho) [inline waspc/waspc.cabal:9]: Let's revert this change as well since we'll be releasing 0.14.0 in mid-July and not 0.14.1 (I'm confident we'll merge this until then)

File: web/versioned_docs/version-0.13.0/auth/Pills.css [hunk 6]
```diff
@@ -1,6 +1,7 @@
 :root {
```
Comment (infomiho) [inline web/versioned_docs/version-0.13.0/auth/Pills.css:1]: Please revert all the changes in the `0.13.0` versioned docs since this feature won't be included in the 0.13.0 version of Wasp.

**Comment by infomiho** [comment]
Great contribution @wardbox 

I'd recommend fixing the e2e tests by running `./run test:e2e:accept-all` and committing the diff. 

I've left a couple of initial comments now, please address them.

I'll be testing this locally and reviewing the code in more detail next.

File: waspc/data/Generator/templates/sdk/wasp/auth/forms/internal/social/SocialIcons.tsx (hunk 2 above)
Comment (infomiho) [reply waspc/data/Generator/templates/sdk/wasp/auth/forms/internal/social/SocialIcons.tsx:17]: Ideally the dimensions would be similar to the rest of the icons
Screenshot 2024-06-24 at 13 34 44

Here's one that works better (from [React SVG Icons](https://reactsvgicons.com/search?q=discord)):
```svg
export const Discord = () => (
  <svg
    className={defaultStyles()}
    aria-hidden="true"
    fill="currentColor"
    viewBox="0 0 16 16"
  >
    <path d="M13.545 2.907a13.227 13.227 0 00-3.257-1.011.05.05 0 00-.052.025c-.141.25-.297.577-.406.833a12.19 12.19 0 00-3.658 0 8.258 8.258 0 00-.412-.833.051.051 0 00-.052-.025c-1.125.194-2.22.534-3.257 1.011a.041.041 0 00-.021.018C.356 6.024-.213 9.047.066 12.032c.001.014.01.028.021.037a13.276 13.276 0 003.995 2.02.05.05 0 00.056-.019c.308-.42.582-.863.818-1.329a.05.05 0 00-.01-.059.051.051 0 00-.018-.011 8.875 8.875 0 01-1.248-.595.05.05 0 01-.02-.066.051.051 0 01.015-.019c.084-.063.168-.129.248-.195a.05.05 0 01.051-.007c2.619 1.196 5.454 1.196 8.041 0a.052.052 0 01.053.007c.08.066.164.132.248.195a.051.051 0 01-.004.085 8.254 8.254 0 01-1.249.594.05.05 0 00-.03.03.052.052 0 00.003.041c.24.465.515.909.817 1.329a.05.05 0 00.056.019 13.235 13.235 0 004.001-2.02.049.049 0 00.021-.037c.334-3.451-.559-6.449-2.366-9.106a.034.034 0 00-.02-.019zm-8.198 7.307c-.789 0-1.438-.724-1.438-1.612 0-.889.637-1.613 1.438-1.613.807 0 1.45.73 1.438 1.613 0 .888-.637 1.612-1.438 1.612zm5.316 0c-.788 0-1.438-.724-1.438-1.612 0-.889.637-1.613 1.438-1.613.807 0 1.451.73 1.438 1.613 0 .888-.631 1.612-1.438 1.612z" />
  </svg>
)
```

File: web/versioned_docs/version-0.13.0/auth/social-auth/discord.md [hunk 7]
```diff
@@ -0,0 +1,549 @@
+---
```
Comment (infomiho) [inline web/versioned_docs/version-0.13.0/auth/social-auth/discord.md:1]: The docs won't build for me:
```
[ERROR] Error: Invalid sidebar file at "sidebars.js".
These sidebar document ids do not exist:
- auth/social-auth/discord

Available document ids are:
- advanced/accessing-app-config
- advanced/apis
- advanced/deployment/cli
- advanced/deployment/manually
- advanced/deployment/overview
- advanced/email/email
- advanced/jobs
- advanced/links
- advanced/middleware-config
- advanced/web-sockets
- auth/auth-hooks
- auth/email
- auth/entities/entities
- auth/overview
- auth/social-auth/github
- auth/social-auth/google
- auth/social-auth/keycloak
- auth/social-auth/overview
- auth/ui
- auth/username-and-pass
- contact
- contributing
- data-model/backends
- data-model/crud
- data-model/entities
- data-model/operations/actions
- data-model/operations/overview
- data-model/operations/queries
- general/cli
- general/language
- introduction/editor-setup
- introduction/introduction
- introduction/quick-start
- migrate-from-0-11-to-0-12
- migrate-from-0-12-to-0-13
- migrate-from-0-13-to-0-14
- project/client-config
- project/css-frameworks
- project/custom-vite-config
- project/customizing-app
- project/dependencies
- project/env-vars
- project/server-config
- project/starter-templates
- project/static-assets
- project/testing
- telemetry
- tutorial/actions
- tutorial/auth
- tutorial/create
- tutorial/entities
- tutorial/pages
- tutorial/project-structure
- tutorial/queries
- vision
- wasp-ai/creating-new-app
- wasp-ai/developing-existing-app
- writingguide
```

This is due to you putting the Discord docs _only_ in the 0.13.0 versioned docs. You need to put them in the `docs/` folder which are the _latest_ docs for the docs to work properly.

Comment (infomiho) [reply web/versioned_docs/version-0.13.0/auth/social-auth/discord.md:1]: Did you try running the docs with `npm run start` in the `web` folder to see them working?

File: web/sidebars.js [hunk 8]
```diff
@@ -68,6 +68,7 @@ module.exports = {
           collapsed: true,
           items: [
             'auth/social-auth/overview',
+            'auth/social-auth/discord',
```
Comment (infomiho) [inline web/sidebars.js:71]: The Discord auth provider shouldn't be the first one in the sidebar as it's not the most popular auth method we offer, let's put in the last place under the social auth group.

File: web/versioned_docs/version-0.13.0/auth/social-auth/discord.md [hunk 9]
```diff
@@ -0,0 +1,549 @@
...
+</TabItem>
+</Tabs>
+
+We imported the generated Auth UI components and used them in our pages. Read more about the Auth UI components [here](../../auth/ui).
+
+### Conclusion
+
+Yay, we've successfully set up Discord Auth! 🎉
+
+![Discord Auth](/img/auth/github.png) <!-- TODO: Grab screenshot of discord auth -->
```
Comment (infomiho) [inline web/versioned_docs/version-0.13.0/auth/social-auth/discord.md:263]: Yep, let's add a Discord auth screenshot 👍

File: web/versioned_docs/version-0.13.0/auth/social-auth/discord.md [hunk 10]
```diff
@@ -0,0 +1,549 @@
...
+  "premium_type": 1,
+  "public_flags": 64,
+  "avatar_decoration_data": {
+    "sku_id": "1144058844004233369",
+    "asset": "a_fed43ab12698df65902ba06727e20c0e"
+  }
+}
+```
+
+The fields you receive will depend on the scopes you requested. The default scope is set to `identify` only. If you want to get the email, you need to specify the `email` scope in the `configFn` function.
```
Comment (infomiho) [inline web/versioned_docs/version-0.13.0/auth/social-auth/discord.md:347]: You say here the default scope it `identity` but in the code we put `identity` and `email`?

File: web/docs/auth/Pills.css (hunk 3 above)
Comment (infomiho) [reply web/docs/auth/Pills.css:4]: We like to be more intentional about the color choices :D Let's go for `#d3d6f2` for light and `#2f3670` for dark variant.

File: web/docs/auth/auth-hooks.md [hunk 11]
```diff
@@ -87,7 +87,7 @@ Wasp calls the `onBeforeSignup` hook before the user is created.
 
 The `onBeforeSignup` hook can be useful if you want to reject a user based on some criteria before they sign up.
 
-Works with <EmailPill /> <UsernameAndPasswordPill /> <GithubPill /> <GooglePill /> <KeycloakPill />
+Works with <EmailPill /> <UsernameAndPasswordPill /> <DiscordPill /> <GithubPill /> <GooglePill /> <KeycloakPill />
```
Comment (infomiho) [inline web/docs/auth/auth-hooks.md:90]: We need to also update the list of pills on the Auth UI page :)

File: web/docs/auth/ui.md [hunk 12]
```diff
@@ -2,7 +2,7 @@
 title: Auth UI
 ---
 
-import { EmailPill, UsernameAndPasswordPill, GithubPill, GooglePill, KeycloakPill } from "./Pills";
+import { EmailPill, UsernameAndPasswordPill, DiscordPill, GithubPill, GooglePill, KeycloakPill } from "./Pills";
```
Comment (infomiho) [inline web/docs/auth/ui.md:5]: Here, we need to use the pill in the correct spots :)

File: waspc/src/Wasp/Generator/AuthProviders.hs [hunk 13]
```diff
@@ -47,10 +47,19 @@ emailAuthProvider =
       E._displayName = "Email and password"
     }
 
+discordAuthProvider :: OA.OAuthAuthProvider
+discordAuthProvider =
+  OA.OAuthAuthProvider
+    { OA._providerId = fromJust $ makeProviderId "discord",
+      OA._displayName = "Discord",
+      OA._requiredScope = ["identify", "email"]
```
[the remark this answers is not in the document]
Comment (wardbox) [reply waspc/src/Wasp/Generator/AuthProviders.hs:55]: Just identify works here, I remember updating docs and not being able to find where I set this - thanks!

File: web/versioned_docs/version-0.13.0/auth/social-auth/discord.md (hunk 7 above)
Comment (wardbox) [reply web/versioned_docs/version-0.13.0/auth/social-auth/discord.md:1]: I did not, wasn't sure how to do that - will update and try this

File: waspc/data/Generator/templates/sdk/wasp/auth/forms/internal/social/SocialIcons.tsx [hunk 14]
```diff
@@ -54,3 +54,14 @@ export const GitHub = () => (
...
+    fill="currentColor"
+    viewBox="0 0 16 16"
+  >
+    <path d="M13.545 2.907a13.227 13.227 0 00-3.257-1.011.05.05 0 00-.052.025c-.141.25-.297.577-.406.833a12.19 12.19 0 00-3.658 0 8.258 8.258 0 00-.412-.833.051.051 0 00-.052-.025c-1.125.194-2.22.534-3.257 1.011a.041.041 0 00-.021.018C.356 6.024-.213 9.047.066 12.032c.001.014.01.028.021.037a13.276 13.276 0 003.995 2.02.05.05 0 00.056-.019c.308-.42.582-.863.818-1.329a.05.05 0 00-.01-.059.051.051 0 00-.018-.011 8.875 8.875 0 01-1.248-.595.05.05 0 01-.02-.066.051.051 0 01.015-.019c.084-.063.168-.129.248-.195a.05.05 0 01.051-.007c2.619 1.196 5.454 1.196 8.041 0a.052.052 0 01.053.007c.08.066.164.132.248.195a.051.051 0 01-.004.085 8.254 8.254 0 01-1.249.594.05.05 0 00-.03.03.052.052 0 00.003.041c.24.465.515.909.817 1.329a.05.05 0 00.056.019 13.235 13.235 0 004.001-2.02.049.049 0 00.021-.037c.334-3.451-.559-6.449-2.366-9.106a.034.034 0 00-.02-.019zm-8.198 7.307c-.789 0-1.438-.724-1.438-1.612 0-.889.637-1.613 1.438-1.613.807 0 1.45.73 1.438 1.613 0 .888-.637 1.612-1.438 1.612zm5.316 0c-.788 0-1.438-.724-1.438-1.612 0-.889.637-1.613 1.438-1.613.807 0 1.451.73 1.438 1.613 0 .888-.631 1.612-1.438 1.612z" />
+  </svg>
+)
```
Comment (infomiho) [inline waspc/data/Generator/templates/sdk/wasp/auth/forms/internal/social/SocialIcons.tsx:67]: Let's add a new line to the end of the file. We strive to do it, here a random gist explain a bit of the reason behind it https://gist.github.com/camh-/1bebfcff1b0f814e9b191edc60d5206b

File: web/versioned_docs/version-0.13.0/auth/Pills.css [hunk 15]
```diff
@@ -14,4 +14,4 @@
   --auth-pills-google: #365314;
   --auth-pills-keycloak: #2d5866;
   --auth-pills-username-and-pass: #831843;
-}
+}
```
Comment (infomiho) [inline web/versioned_docs/version-0.13.0/auth/Pills.css:17]: Let's also add here a newline :)

File: web/versioned_sidebars/version-0.13.0-sidebars.json [hunk 16]
```diff
@@ -149,4 +149,4 @@
       ]
     }
   ]
-}
+}
```
Comment (infomiho) [inline web/versioned_sidebars/version-0.13.0-sidebars.json:152]: Let's also add here a newline :)

**Review by infomiho — COMMENTED** [review]
Thanks once more for the contribution. We are almost there! 

I left a couple of minor comments regarding newlines and I think we still need to update the pill colors in the docs. I've tested it locally, works great.

**Review by infomiho — APPROVED** [review]
